File: aws-examples/elixir/minimal-s3-demo/test/s3_waiters_test.exs
defmodule S3WaitersTest do
  use ExUnit.Case, async: true

  alias AmazonS3Types.{HeadBucketInput, HeadBucketOutput}

  @bucket "my-bucket"
  @wait_opts [max_attempts: 3, min_delay_ms: 0, max_delay_ms: 0]

  setup do
    on_exit(fn -> :persistent_term.erase({HeadBucketHttpMock, :responses}) end)
    on_exit(fn -> :persistent_term.erase({HeadBucketHttpMock, :count}) end)
    :ok
  end

  test "wait_bucket_exists/3 returns ok when head_bucket succeeds" do
    HeadBucketHttpMock.reset([{:status, 200}])

    assert {:ok, {:ok, %HeadBucketOutput{}}} =
             AmazonS3Waiters.wait_bucket_exists(client_config(), head_bucket_input(), @wait_opts)

    assert HeadBucketHttpMock.call_count() == 1
  end

  test "wait_bucket_exists/3 retries until head_bucket succeeds" do
    HeadBucketHttpMock.reset([{:status, 404}, {:status, 200}])

    assert {:ok, {:ok, %HeadBucketOutput{}}} =
             AmazonS3Waiters.wait_bucket_exists(client_config(), head_bucket_input(), @wait_opts)

    assert HeadBucketHttpMock.call_count() == 2
  end

  test "wait_bucket_exists/3 returns max_attempts_exceeded when bucket never appears" do
    HeadBucketHttpMock.reset([{:status, 404}, {:status, 404}, {:status, 404}])

    assert {:error, :max_attempts_exceeded} =
             AmazonS3Waiters.wait_bucket_exists(
               client_config(),
               head_bucket_input(),
               Keyword.put(@wait_opts, :max_attempts, 2)
             )

    assert HeadBucketHttpMock.call_count() == 2
  end

  test "wait_bucket_not_exists/3 returns max_attempts_exceeded when bucket still exists" do
    HeadBucketHttpMock.reset([{:status, 200}, {:status, 200}])

    assert {:error, :max_attempts_exceeded} =
             AmazonS3Waiters.wait_bucket_not_exists(
               client_config(),
               head_bucket_input(),
               Keyword.put(@wait_opts, :max_attempts, 2)
             )

    assert HeadBucketHttpMock.call_count() == 2
  end

  defp client_config do
    %{
      base_url: "http://localhost:4566",
      region: "us-east-1",
      endpoint_prefix: "s3",
      signing_name: "s3",
      s3_addressing_style: :path_style,
      http_client: HeadBucketHttpMock,
      credentials: nil,
      retry: [max_attempts: 1, base_delay_ms: 0]
    }
  end

  defp head_bucket_input do
    %HeadBucketInput{bucket: @bucket}
  end
end

File: aws-examples/erlang/minimal-s3-demo/test/s3_waiters_test.erl
-module(s3_waiters_test).

-include_lib("eunit/include/eunit.hrl").
-include("amazon_s3_types.hrl").

-define(BUCKET, <<"my-bucket">>).
-define(WAIT_OPTS, #{max_attempts => 3, min_delay_ms => 0, max_delay_ms => 0}).

wait_bucket_exists_ok_test() ->
    with_cleanup(fun() ->
        head_bucket_http_mock:reset([{status, 200}]),
        Input = #head_bucket_input{bucket = ?BUCKET},
        ?assertMatch(
            {ok, {ok, #head_bucket_output{}}},
            amazon_s3_waiters:wait_bucket_exists(client_config(), Input, ?WAIT_OPTS)
        ),
        ?assertEqual(1, head_bucket_http_mock:call_count())
    end).

wait_bucket_exists_retries_test() ->
    with_cleanup(fun() ->
        head_bucket_http_mock:reset([{status, 404}, {status, 200}]),
        Input = #head_bucket_input{bucket = ?BUCKET},
        ?assertMatch(
            {ok, {ok, #head_bucket_output{}}},
            amazon_s3_waiters:wait_bucket_exists(client_config(), Input, ?WAIT_OPTS)
        ),
        ?assertEqual(2, head_bucket_http_mock:call_count())
    end).

wait_bucket_exists_max_attempts_test() ->
    with_cleanup(fun() ->
        head_bucket_http_mock:reset([{status, 404}, {status, 404}, {status, 404}]),
        Input = #head_bucket_input{bucket = ?BUCKET},
        Opts = maps:put(max_attempts, 2, ?WAIT_OPTS),
        ?assertEqual(
            {error, max_attempts_exceeded},
            amazon_s3_waiters:wait_bucket_exists(client_config(), Input, Opts)
        ),
        ?assertEqual(2, head_bucket_http_mock:call_count())
    end).

wait_bucket_not_exists_max_attempts_test() ->
    with_cleanup(fun() ->
        head_bucket_http_mock:reset([{status, 200}, {status, 200}]),
        Input = #head_bucket_input{bucket = ?BUCKET},
        Opts = maps:put(max_attempts, 2, ?WAIT_OPTS),
        ?assertEqual(
            {error, max_attempts_exceeded},
            amazon_s3_waiters:wait_bucket_not_exists(client_config(), Input, Opts)
        ),
        ?assertEqual(2, head_bucket_http_mock:call_count())
    end).

client_config() ->
    #{
        base_url => <<"http://localhost:4566">>,
        region => <<"us-east-1">>,
        endpoint_prefix => <<"s3">>,
        signing_name => <<"s3">>,
        s3_addressing_style => path_style,
        http_client => head_bucket_http_mock,
        credentials => undefined,
        retry => #{max_attempts => 1, base_delay_ms => 0}
    }.

with_cleanup(Fun) ->
    try
        Fun()
    after
        persistent_term:erase({head_bucket_http_mock, responses}),
        persistent_term:erase({head_bucket_http_mock, count})
    end.
